File: backend/src/test/java/com/joying/chat/service/ChatRoomPermissionCacheTest.java
package com.joying.chat.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import java.util.Optional;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.mockito.junit.jupiter.MockitoSettings;
import org.mockito.quality.Strictness;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.ValueOperations;

import com.joying.chat.domain.ChatRoom;
import com.joying.chat.domain.ChatRoomMember;
import com.joying.chat.domain.ChatRoomStatus;
import com.joying.chat.repository.ChatRoomMemberRepository;
import com.joying.chat.repository.ChatRoomRepository;
import com.joying.member.domain.Member;

/**
 * 채팅방 권한 캐시의 현재 동작을 고정한다.
 *
 * <p>실사에서 결함이 나온 자리라 특히 조심해서 박아 둔다. 캐시가 무엇을 근거로
 * 판정하는지, 무엇을 보지 않는지를 테스트가 그대로 드러내게 썼다.
 */
@ExtendWith(MockitoExtension.class)
@MockitoSettings(strictness = Strictness.LENIENT)
class ChatRoomPermissionCacheTest {

	private static final long ROOM_ID = 1L;
	private static final long BUYER_ID = 100L;
	private static final long SELLER_ID = 200L;
	private static final long STRANGER_ID = 999L;

	@Mock
	RedisTemplate<String, String> redis;

	@Mock
	ValueOperations<String, String> valueOps;

	@Mock
	ChatRoomRepository chatRoomRepository;

	@Mock
	ChatRoomMemberRepository chatRoomMemberRepository;

	ChatRoomPermissionCache cache;

	@BeforeEach
	void setUp() {
		given(redis.opsForValue()).willReturn(valueOps);
		cache = new ChatRoomPermissionCache(redis, chatRoomRepository, chatRoomMemberRepository);
	}

	/** 방에 남아 있는 참여자 */
	private void memberStaying(long memberId) {
		ChatRoomMember member = mock(ChatRoomMember.class);
		given(member.isLeft()).willReturn(false);
		given(chatRoomMemberRepository.findByChatRoomIdAndMemberId(ROOM_ID, memberId))
			.willReturn(Optional.of(member));
	}

	/** 방을 나간 참여자 */
	private void memberLeft(long memberId) {
		ChatRoomMember member = mock(ChatRoomMember.class);
		given(member.isLeft()).willReturn(true);
		given(chatRoomMemberRepository.findByChatRoomIdAndMemberId(ROOM_ID, memberId))
			.willReturn(Optional.of(member));
	}

	private String key(long memberId) {
		return "chatroom:permission:" + ROOM_ID + ":" + memberId;
	}

	private ChatRoom roomWith(long buyerId, long sellerId) {
		Member buyer = mock(Member.class);
		Member seller = mock(Member.class);
		given(buyer.getMemberId()).willReturn(buyerId);
		given(seller.getMemberId()).willReturn(sellerId);

		ChatRoom room = mock(ChatRoom.class);
		given(room.getBuyer()).willReturn(buyer);
		given(room.getSeller()).willReturn(seller);
		given(room.getStatus()).willReturn(ChatRoomStatus.ACTIVE);
		return room;
	}

	@Test
	@DisplayName("캐시에 ALLOWED가 있으면 저장소를 보지 않는다")
	void cacheHitSkipsDatabase() {
		given(valueOps.get(key(BUYER_ID))).willReturn("ALLOWED");

		assertThat(cache.hasPermission(ROOM_ID, BUYER_ID)).isTrue();
		verify(chatRoomRepository, never()).findById(any());
	}

	@Test
	@DisplayName("캐시에 DENIED가 있으면 거절이다")
	void cachedDenialIsHonored() {
		given(valueOps.get(key(STRANGER_ID))).willReturn("DENIED");

		assertThat(cache.hasPermission(ROOM_ID, STRANGER_ID)).isFalse();
		verify(chatRoomRepository, never()).findById(any());
	}

	@Test
	@DisplayName("캐시가 비었으면 저장소를 보고 결과를 캐싱한다")
	void cacheMissChecksDatabaseAndCaches() {
		ChatRoom room = roomWith(BUYER_ID, SELLER_ID);
		given(valueOps.get(key(BUYER_ID))).willReturn(null);
		given(chatRoomRepository.findById(ROOM_ID)).willReturn(Optional.of(room));
		memberStaying(BUYER_ID);

		assertThat(cache.hasPermission(ROOM_ID, BUYER_ID)).isTrue();
		verify(valueOps).set(eq(key(BUYER_ID)), eq("ALLOWED"), anyLong(), eq(TimeUnit.HOURS));
	}

	@Test
	@DisplayName("방에 없는 사람은 거절이고 그것도 캐싱한다")
	void strangerIsDeniedAndCached() {
		ChatRoom room = roomWith(BUYER_ID, SELLER_ID);
		given(valueOps.get(key(STRANGER_ID))).willReturn(null);
		given(chatRoomRepository.findById(ROOM_ID)).willReturn(Optional.of(room));

		assertThat(cache.hasPermission(ROOM_ID, STRANGER_ID)).isFalse();
		verify(valueOps).set(eq(key(STRANGER_ID)), eq("DENIED"), anyLong(), eq(TimeUnit.HOURS));
	}

	@Test
	@DisplayName("방이 없으면 거절이다")
	void missingRoomIsDenied() {
		given(valueOps.get(key(BUYER_ID))).willReturn(null);
		given(chatRoomRepository.findById(ROOM_ID)).willReturn(Optional.empty());

		assertThat(cache.hasPermission(ROOM_ID, BUYER_ID)).isFalse();
	}

	@Test
	@DisplayName("방을 나간 사람은 거절한다. 무효화가 결과를 바꿀 수 있어야 한다")
	void deniesMemberWhoLeft() {
		// 예전에는 방의 구매자이거나 판매자인가만 봤다. 그래서 나가기에서 캐시를
		// 지워도 다음 조회가 같은 근거로 다시 허용을 계산해 캐싱했다. 무효화가
		// 결과를 바꿀 수 없다는 뜻이었다.
		ChatRoom room = roomWith(BUYER_ID, SELLER_ID);
		given(valueOps.get(key(BUYER_ID))).willReturn(null);
		given(chatRoomRepository.findById(ROOM_ID)).willReturn(Optional.of(room));
		memberLeft(BUYER_ID);

		assertThat(cache.hasPermission(ROOM_ID, BUYER_ID)).isFalse();
		verify(valueOps).set(eq(key(BUYER_ID)), eq("DENIED"), anyLong(), eq(TimeUnit.HOURS));
	}

	@Test
	@DisplayName("참여자 기록이 없으면 거절한다")
	void deniesWhenNoMembershipRecord() {
		ChatRoom room = roomWith(BUYER_ID, SELLER_ID);
		given(valueOps.get(key(BUYER_ID))).willReturn(null);
		given(chatRoomRepository.findById(ROOM_ID)).willReturn(Optional.of(room));
		given(chatRoomMemberRepository.findByChatRoomIdAndMemberId(ROOM_ID, BUYER_ID))
			.willReturn(Optional.empty());

		assertThat(cache.hasPermission(ROOM_ID, BUYER_ID)).isFalse();
	}

	@Test
	@DisplayName("닫힌 방은 거절한다")
	void deniesClosedRoom() {
		ChatRoom room = roomWith(BUYER_ID, SELLER_ID);
		given(room.getStatus()).willReturn(ChatRoomStatus.CLOSED);
		given(valueOps.get(key(BUYER_ID))).willReturn(null);
		given(chatRoomRepository.findById(ROOM_ID)).willReturn(Optional.of(room));

		assertThat(cache.hasPermission(ROOM_ID, BUYER_ID)).isFalse();
	}
}

File: backend/src/test/kotlin/com/joying/chat/service/ChatRoomPermissionCacheTest.kt
package com.joying.chat.service

import com.joying.chat.domain.ChatRoom
import com.joying.chat.repository.ChatRoomRepository
import com.joying.member.domain.Member
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.core.ValueOperations
import java.util.Optional
import java.util.concurrent.TimeUnit

/**
 * 채팅방 권한 캐시의 현재 동작을 고정한다.
 *
 * 실사에서 결함이 나온 자리라 특히 조심해서 박아 둔다. 캐시가 무엇을 근거로
 * 판정하는지, 무엇을 보지 않는지를 테스트가 그대로 드러내게 썼다.
 */
class ChatRoomPermissionCacheTest {

    private lateinit var redis: RedisTemplate<String, String>
    private lateinit var valueOps: ValueOperations<String, String>
    private lateinit var chatRoomRepository: ChatRoomRepository
    private lateinit var cache: ChatRoomPermissionCache

    private val roomId = 1L
    private val buyerId = 100L
    private val sellerId = 200L
    private val strangerId = 999L

    @BeforeEach
    fun setUp() {
        redis = mock()
        valueOps = mock()
        chatRoomRepository = mock()
        whenever(redis.opsForValue()).thenReturn(valueOps)
        cache = ChatRoomPermissionCache(redis, chatRoomRepository)
    }

    private fun key(memberId: Long) = "chatroom:permission:$roomId:$memberId"

    private fun roomWith(buyer: Long, seller: Long): ChatRoom {
        val buyerMember = mock<Member>()
        val sellerMember = mock<Member>()
        whenever(buyerMember.getMemberId()).thenReturn(buyer)
        whenever(sellerMember.getMemberId()).thenReturn(seller)

        val room = mock<ChatRoom>()
        whenever(room.buyer).thenReturn(buyerMember)
        whenever(room.seller).thenReturn(sellerMember)
        return room
    }

    @Test
    @DisplayName("캐시에 ALLOWED가 있으면 저장소를 보지 않는다")
    fun cacheHitSkipsDatabase() = runBlocking<Unit> {
        whenever(valueOps.get(key(buyerId))).thenReturn("ALLOWED")

        assertThat(cache.hasPermission(roomId, buyerId)).isTrue()
        verify(chatRoomRepository, never()).findById(any())
    }

    @Test
    @DisplayName("캐시에 DENIED가 있으면 거절이다")
    fun cachedDenialIsHonored() = runBlocking<Unit> {
        whenever(valueOps.get(key(strangerId))).thenReturn("DENIED")

        assertThat(cache.hasPermission(roomId, strangerId)).isFalse()
        verify(chatRoomRepository, never()).findById(any())
    }

    @Test
    @DisplayName("캐시가 비었으면 저장소를 보고 결과를 캐싱한다")
    fun cacheMissChecksDatabaseAndCaches() = runBlocking<Unit> {
        val room = roomWith(buyerId, sellerId)
        whenever(valueOps.get(key(buyerId))).thenReturn(null)
        whenever(chatRoomRepository.findById(roomId)).thenReturn(Optional.of(room))

        assertThat(cache.hasPermission(roomId, buyerId)).isTrue()
        verify(valueOps).set(eq(key(buyerId)), eq("ALLOWED"), any(), eq(TimeUnit.HOURS))
    }

    @Test
    @DisplayName("방에 없는 사람은 거절이고 그것도 캐싱한다")
    fun strangerIsDeniedAndCached() = runBlocking<Unit> {
        val room = roomWith(buyerId, sellerId)
        whenever(valueOps.get(key(strangerId))).thenReturn(null)
        whenever(chatRoomRepository.findById(roomId)).thenReturn(Optional.of(room))

        assertThat(cache.hasPermission(roomId, strangerId)).isFalse()
        verify(valueOps).set(eq(key(strangerId)), eq("DENIED"), any(), eq(TimeUnit.HOURS))
    }

    @Test
    @DisplayName("방이 없으면 거절이다")
    fun missingRoomIsDenied() = runBlocking<Unit> {
        whenever(valueOps.get(key(buyerId))).thenReturn(null)
        whenever(chatRoomRepository.findById(roomId)).thenReturn(Optional.empty())

        assertThat(cache.hasPermission(roomId, buyerId)).isFalse()
    }

    @Test
    @DisplayName("알려진 결함: 판정 근거가 buyer/seller뿐이라 방을 나갔는지는 보지 않는다")
    fun knownDefect_ignoresWhetherMemberLeft() = runBlocking<Unit> {
        // 판정은 chatRoom의 buyer/seller와 같은 사람인가만 본다. ChatRoomMember의
        // isLeft도, 방의 status도 보지 않는다.
        //
        // 그래서 나가기에서 캐시를 지워도 다음 조회가 같은 근거로 다시 ALLOWED를
        // 계산해 캐싱한다. 무효화가 결과를 바꿀 수 없다는 뜻이다.
        //
        // 나간 사람을 실제로 막는 것은 이 캐시가 아니라 메시지 전송 경로의 별도
        // 조회다. 조회 경로에는 그 검사가 없어 히스토리가 열려 있다.
        //
        // 이관 중에 고치지 않는다. 지금 동작을 박아 두고 이관이 끝난 뒤 뒤집는다.
        val room = roomWith(buyerId, sellerId)
        whenever(valueOps.get(key(buyerId))).thenReturn(null)
        whenever(chatRoomRepository.findById(roomId)).thenReturn(Optional.of(room))

        assertThat(cache.hasPermission(roomId, buyerId))
            .`as`("나갔더라도 buyer이면 통과한다")
            .isTrue()
    }
}
